### PR TITLE
feat!: support generic inputs and outputs in controllers

### PR DESCRIPTION
**Pull request by bpapaspyros**
## Description

This PR adds support for generic outputs in controllers. Similarly to #145, we use compile-time data type inference to make sure that custom (ROS) types are 1) added to the output map generically, 2) are directly published when `write_output` is invoked.

- #151 

## Review guidelines

Estimated Time of Review: 10 minutes

## Checklist before merging:

- [ ] Confirm that the relevant changelog(s) are up-to-date in case of any user-facing changes

### CODE REVIEW

File: source/modulo_controllers/test/test_controller_interface.cpp [hunk 1]
```diff
@@ -186,7 +188,23 @@ TYPED_TEST_P(ControllerInterfaceTest, OutputTest) {
   }
 }
 
-REGISTER_TYPED_TEST_CASE_P(ControllerInterfaceTest, ConfigureErrorTest, InputTest, OutputTest);
+TYPED_TEST_P(ControllerInterfaceTest, CustomOutputTest) {
```
Comment (bpapaspyros) [inline source/modulo_controllers/test/test_controller_interface.cpp:191]: If the rest of the PR makes sense, I think it would be reasonable that we merge this test with the above function.

File: source/modulo_controllers/include/modulo_controllers/BaseControllerInterface.hpp [hunk 2]
```diff
@@ -569,8 +577,25 @@ BaseControllerInterface::create_subscription(const std::string& name, const std:
...
 inline void BaseControllerInterface::add_output(const std::string& name, const std::string& topic_name) {
-  std::shared_ptr<state_representation::State> state_ptr = std::make_shared<T>();
-  create_output(EncodedStatePublishers(state_ptr, {}, {}), name, topic_name);
+  if constexpr (modulo_core::concepts::CustomT<T>) {
+    typedef std::pair<std::shared_ptr<rclcpp::Publisher<T>>, realtime_tools::RealtimePublisherSharedPtr<T>> PublisherT;
+    auto parsed_name = validate_and_declare_signal(name, "output", topic_name);
+    if (!parsed_name.empty()) {
+      outputs_.insert_or_assign(name, PublisherT());
+      custom_output_configuration_callables_.insert_or_assign(
+          name, [this](CustomPublishers& pub, const std::string& name) {
```
Comment (domire8) [inline source/modulo_controllers/include/modulo_controllers/BaseControllerInterface.hpp:586]:
```suggestion
      outputs_.insert_or_assign(parsed_name, PublisherT());
      custom_output_configuration_callables_.insert_or_assign(
          parsed_name, [this](CustomPublishers& pub, const std::string& name) {
```

Comment (domire8) [reply source/modulo_controllers/include/modulo_controllers/BaseControllerInterface.hpp:586]: And also correct that in the `create_output` please

File: source/modulo_controllers/include/modulo_controllers/BaseControllerInterface.hpp [hunk 3]
```diff
@@ -689,44 +714,84 @@ inline std::optional<std::string> BaseControllerInterface::read_input<std::strin
...
-    return;
-  }
-  auto rt_pub = std::get<2>(publishers);
-  if (rt_pub && rt_pub->trylock()) {
+  if constexpr (modulo_core::concepts::CustomT<T>) {
+    if (outputs_.find(name) == outputs_.end()) {
+      RCLCPP_WARN_THROTTLE(
+          get_node()->get_logger(), *get_node()->get_clock(), 1000, "Could not find output '%s'", name.c_str());
+      return;
+    }
```
Comment (domire8) [inline source/modulo_controllers/include/modulo_controllers/BaseControllerInterface.hpp:722]: you can move this out of the conditional since it's repeated further down

File: source/modulo_controllers/include/modulo_controllers/BaseControllerInterface.hpp [hunk 4]
```diff
@@ -689,44 +714,84 @@ inline std::optional<std::string> BaseControllerInterface::read_input<std::strin
...
-          get_node()->get_logger(), *get_node()->get_clock(), 1000, "Failed to publish output '%s': %s", name.c_str(),
-          ex.what());
+          get_node()->get_logger(), *get_node()->get_clock(), 1000,
+          "Skipping publication of output '%s' due to wrong data type: %s", name.c_str(), ex.what());
+    }
+
+    if (rt_pub && rt_pub->trylock()) {
+      try {
+        rt_pub->msg_ = data;
+      } catch (const modulo_core::exceptions::MessageTranslationException& ex) {
```
Comment (domire8) [inline source/modulo_controllers/include/modulo_controllers/BaseControllerInterface.hpp:748]: This exception doesn't make sense here

File: source/modulo_controllers/test/test_controller_interface.cpp (hunk 1 above)
Comment (domire8) [reply source/modulo_controllers/test/test_controller_interface.cpp:191]: If possible then yes

File: source/modulo_controllers/include/modulo_controllers/BaseControllerInterface.hpp (hunk 4 above)
Comment (bpapaspyros) [reply source/modulo_controllers/include/modulo_controllers/BaseControllerInterface.hpp:748]: oups, yes

File: source/modulo_controllers/include/modulo_controllers/BaseControllerInterface.hpp [hunk 5]
```diff
@@ -569,8 +577,25 @@ BaseControllerInterface::create_subscription(const std::string& name, const std:
...
 inline void BaseControllerInterface::add_output(const std::string& name, const std::string& topic_name) {
-  std::shared_ptr<state_representation::State> state_ptr = std::make_shared<T>();
-  create_output(EncodedStatePublishers(state_ptr, {}, {}), name, topic_name);
+  if constexpr (modulo_core::concepts::CustomT<T>) {
+    typedef std::pair<std::shared_ptr<rclcpp::Publisher<T>>, realtime_tools::RealtimePublisherSharedPtr<T>> PublisherT;
+    auto parsed_name = validate_and_declare_signal(name, "output", topic_name);
+    if (!parsed_name.empty()) {
+      outputs_.insert_or_assign(parsed_name, PublisherT());
+      custom_output_configuration_callables_.insert_or_assign(
+          name, [this](CustomPublishers& pub, const std::string& name) {
```
Comment (domire8) [inline source/modulo_controllers/include/modulo_controllers/BaseControllerInterface.hpp:586]:
```suggestion
          name, [this](CustomPublishers& pub, const std::string& topic) {
```

File: source/modulo_controllers/test/test_controller_interface.cpp (hunk 1 above)
Comment (bpapaspyros) [reply source/modulo_controllers/test/test_controller_interface.cpp:191]: I'll spend a quick moment to see if I can make the tests a bit nicer, but otherwise the ones I added already test enough.

File: source/modulo_controllers/src/BaseControllerInterface.cpp [hunk 6]
```diff
@@ -403,10 +406,15 @@ void BaseControllerInterface::add_outputs() {
               [&](StringPublishers& pub) {
                 pub.first = get_node()->create_publisher<std_msgs::msg::String>(topic, qos_);
                 pub.second = std::make_shared<realtime_tools::RealtimePublisher<std_msgs::msg::String>>(pub.first);
+              },
+              [&](CustomPublishers& pub) {
+                custom_output_configuration_callables_.at(name)(pub, name);
               }},
           publishers);
+    } catch (const std::bad_any_cast& ex) {
```
Comment (domire8) [inline source/modulo_controllers/src/BaseControllerInterface.cpp:414]: Do you want to do the same catch in `add_inputs`?

Comment (bpapaspyros) [reply source/modulo_controllers/src/BaseControllerInterface.cpp:414]: the thing is in `add_inputs` we call the `custom_input_configuration_callables` lamda which the only thing it does is to `push_back` another lambda to `subscriptions`. It's only that lambda that does `std::any_cast`. That's why my original question was about adding a try catch in that lamda.

Now, it could be that someone does `add_input<TypeA>` and then does `read_input<TypeB>`, at which point `read_input` would throw a `bad_any_cast` that is handled. The only way for the subscription lamda to throw would be if a message of different type made it all the way to it. But I think ROS would already complain on its own if a message of `TypeB` is published onto a topic of a different `TypeA`

Long story short, I think we might be safe to not add something more here. Does this sound right to you?

Comment (domire8) [reply source/modulo_controllers/src/BaseControllerInterface.cpp:414]: Yeah sounds good :+1:

**Review by domire8 — APPROVED** [review]
:rocket:
